core/src/main/kotlin/org/opensearch/alerting/core/lock/LockService.kt

```diff
@@ -0,0 +1,311 @@
+import org.opensearch.action.delete.DeleteRequest
+import org.opensearch.action.delete.DeleteResponse
+import org.opensearch.action.get.GetRequest
+import org.opensearch.action.get.GetResponse
+import org.opensearch.action.index.IndexRequest
+import org.opensearch.action.index.IndexResponse
+import org.opensearch.action.update.UpdateRequest
+import org.opensearch.action.update.UpdateResponse
+import org.opensearch.client.Client
+import org.opensearch.cluster.service.ClusterService
+import org.opensearch.common.settings.Settings
+import org.opensearch.common.xcontent.LoggingDeprecationHandler
+import org.opensearch.common.xcontent.XContentFactory
+import org.opensearch.common.xcontent.XContentType
+import org.opensearch.commons.alerting.model.ScheduledJob
+import org.opensearch.core.action.ActionListener
+import org.opensearch.core.xcontent.NamedXContentRegistry
+import org.opensearch.core.xcontent.ToXContent
+import org.opensearch.index.IndexNotFoundException
+import org.opensearch.index.engine.DocumentMissingException
+import org.opensearch.index.engine.VersionConflictEngineException
+import org.opensearch.index.seqno.SequenceNumbers
+import java.io.IOException
+import java.time.Instant
+
+private val log = LogManager.getLogger(LockService::class.java)
+
+class LockService(private val client: Client, private val clusterService: ClusterService) {
+    private var testInstant: Instant? = null
+
+    companion object {
+        const val LOCK_INDEX_NAME = ".opensearch-alerting-config-lock"
+
+        @JvmStatic
+        fun lockMapping(): String? {
+            return LockService::class.java.classLoader.getResource("mappings/opensearch-alerting-config-lock.json")
+                ?.readText()
+        }
+    }
+
+    fun lockIndexExist(): Boolean {
+        return clusterService.state().routingTable().hasIndex(LOCK_INDEX_NAME)
+    }
+
+    fun acquireLock(
+        scheduledJob: ScheduledJob,
+        listener: ActionListener<LockModel?>
+    ) {
+        val scheduledJobId = scheduledJob.id
+        acquireLockWithId(scheduledJobId, listener)
+    }
+
+    fun acquireLockWithId(
+        scheduledJobId: String,
+        listener: ActionListener<LockModel?>
+    ) {
+        val lockId = LockModel.generateLockId(scheduledJobId)
+        createLockIndex(
+            object : ActionListener<Boolean> {
+                override fun onResponse(created: Boolean) {
+                    if (created) {
+                        try {
+                            findLock(
+                                lockId,
+                                object : ActionListener<LockModel> {
+                                    override fun onResponse(existingLock: LockModel?) {
+                                        if (existingLock != null) {
+                                            if (isLockReleased(existingLock)) {
+                                                log.debug("lock is released or expired: {}", existingLock)
```

Comment:
Non-blocking - this line is 10x nested. We should avoid deeply nesting code as it hurts readability. We can break much of this into private methods to avoid the nesting